File: web/vuejs/src/components/shared/padding.ts
/**
 * Copyright (c) 2025 worldiety GmbH
 *
 * This file is part of the NAGO Low-Code Platform.
 * Licensed under the terms specified in the LICENSE file.
 *
 * SPDX-License-Identifier: Custom-License
 */
import { cssLengthValue } from '@/components/shared/length';
import { Padding } from '@/shared/proto/nprotoc_gen';

// paddingCSS applies the padding length values. Note, that negative paddings are interpreted as negative margins,
// because negative padding values are not allowed and it seems practical to move views around for some nice effects.
export function paddingCSS(padding?: Padding): string[] {
	const styles: string[] = [];

	if (!padding) {
		return styles;
	}

	if (padding.bottom) {
		if (padding.bottom.startsWith('-')) {
			styles.push(`margin-bottom:${cssLengthValue(padding.bottom)}`);
		} else {
			styles.push(`padding-bottom:${cssLengthValue(padding.bottom)}`);
		}
	}

	if (padding.top) {
		if (padding.top.startsWith('-')) {
			styles.push(`margin-top: ${cssLengthValue(padding.top)}`);
		} else {
			styles.push(`padding-top:${cssLengthValue(padding.top)}`);
		}
	}

	if (padding.right) {
		if (padding.right.startsWith('-')) {
			styles.push(`margin-right:${cssLengthValue(padding.right)}`);
		} else {
			styles.push(`padding-right:${cssLengthValue(padding.right)}`);
		}
	}

	if (padding.left) {
		if (padding.left.startsWith('-')) {
			styles.push(`margin-left:${cssLengthValue(padding.left)}`);
		} else {
			styles.push(`padding-left:${cssLengthValue(padding.left)}`);
		}
	}

	return styles;
}

// marginCSS is like padding but interprets all padding lengths as margin length
export function marginCSS(padding?: Padding): string[] {
	const styles: string[] = [];

	if (!padding) {
		return styles;
	}

	if (padding.bottom) {
		styles.push(`margin-bottom:${cssLengthValue(padding.bottom)}`);
	}

	if (padding.top) {
		styles.push(`margin-top: ${cssLengthValue(padding.top)}`);
	}

	if (padding.right) {
		styles.push(`margin-right:${cssLengthValue(padding.right)}`);
	}

	if (padding.left) {
		styles.push(`margin-left:${cssLengthValue(padding.left)}`);
	}

	return styles;
}

File: presentation/ui2/padding.go
package ui

import "go.wdy.de/nago/presentation/ora"

type Padding struct {
	Top    Length
	Left   Length
	Right  Length
	Bottom Length
}

func (p Padding) ora() ora.Padding {
	return ora.Padding{
		Top:    ora.Length(p.Top),
		Left:   ora.Length(p.Left),
		Right:  ora.Length(p.Right),
		Bottom: ora.Length(p.Bottom),
	}
}

func (p Padding) All(pad Length) Padding {
	p.Left = pad
	p.Right = pad
	p.Bottom = pad
	p.Top = pad
	return p
}

// Vertical means Y axis, so top and bottom are set to the padding value.
func (p Padding) Vertical(pad Length) Padding {
	p.Bottom = pad
	p.Top = pad
	return p
}

// Horizontal means X axis, so left and right are set to the padding value.
func (p Padding) Horizontal(pad Length) Padding {
	p.Left = pad
	p.Right = pad
	return p
}
